src/util/bson/bson.cpp

```diff
@@ -0,0 +1,1172 @@
+
+    bool string(string_t& val) override;
+
+    bool key(string_t& val) override;
+
+    bool start_object(std::size_t) override;
+
+    bool end_object() override;
+
+    bool start_array(std::size_t) override;
+
+    bool end_array() override;
+
+    bool parse_error(std::size_t,
+                     const std::string&,
+                     const nlohmann::detail::exception& ex) override;
+
+    Bson parse(const std::string& json);
+
+protected:
+    struct Instruction {
+        State type;
+        std::string key;
+    };
+
+    /// Convenience class that overloads similar collection functionality
+    class BsonContainer {
+        typedef enum Type {
+            DOCUMENT, ARRAY
+        } Type;
+        Type m_type;
+
+        union {
+            BsonDocument* document;
+            BsonArray* array;
+        };
+    public:
+        ~BsonContainer() noexcept
+        {
+            if (m_type == DOCUMENT) {
+                delete document;
+                document = nullptr;
+            } else {
+                delete array;
+                array = nullptr;
+            }
+        }
+
+        BsonContainer(const BsonDocument& v) noexcept : document(new BsonDocument(v)) {
+            m_type = DOCUMENT;
+        }
+        BsonContainer(const BsonArray& v) noexcept : array(new BsonArray(v)) {
+            m_type = ARRAY;
+        }
+        BsonContainer(BsonDocument&& v) noexcept :
+            m_type(DOCUMENT), document(new BsonDocument(std::move(v))) {}
+
+        BsonContainer(BsonArray&& v) noexcept : m_type(ARRAY), array(new BsonArray(std::move(v))) {}
+
+        explicit operator const BsonDocument&&() const {
+            REALM_ASSERT(is_document());
+            return std::move(*document);
+        }
+
+        explicit operator const BsonArray&&() const {
+            REALM_ASSERT(is_array());
+            return std::move(*array);
+        }
+
+        BsonContainer(const BsonContainer& v)
+        {
+            m_type = v.m_type;
+            if (m_type == DOCUMENT) {
+                document = new BsonDocument;
+                *document = *v.document;
+            } else {
+                array = new BsonArray;
+                *array = *v.array;
+            }
+        }
+
+        BsonContainer& operator=(const BsonContainer& v)
+        {
+            if (&v == this) return *this;
+            this->~BsonContainer();
+
+            m_type = v.m_type;
+            if (m_type == DOCUMENT) {
+                document = new BsonDocument;
+                *document = *v.document;
+            } else {
+                array = new BsonArray;
+                *array = *v.array;
+            }
+            return *this;
+        }
+
+        BsonContainer& operator=(BsonContainer&& v)
+        {
+            if (&v == this) return *this;
+            this->~BsonContainer();
+
+            m_type = v.m_type;
+            if (m_type == DOCUMENT) {
+                if (document) delete document;
```

Comment:
the remaining deletes in this method become redundant because things have already been cleaned up earlier (they are harmless because you did set things to nullptr, but for readability)